File: 647-Palindromic Substrings.py
# !/usr/bin/env python3
# -*- coding:utf-8 -*-
# created on 20170809
# 列表生成式+"DP"+逆序string


class Solution(object):
    def isPlaindromic(self,s):
        # if s==s[::-1]:
        #     return 1
        # else:
        #     return 0
        return s==s[::-1]

    def countSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        """
        l=len(s)
        dp=[[]for i in range(l+1)]
        dp[0]=0
        for i in range(1,l+1):
            dp[i] = dp[i-1]+sum(self.isPlaindromic(s[j:i]) for j in range(i))
        return dp[l]

s = Solution()
print(s.countSubstrings("aba"))

File: 647-Palindromic Substrings.cpp
#include <iostream>
#include <vector>
using namespace std;

//20170809
//DP
class Solution {
public:
    int countSubstrings(string s) {
        int l=s.size();
        vector<vector<int>> dp(l,vector<int>(l,0));
        int r=0;
        for(int i=0;i<l;i++){
            dp[i][i]=1;
            for(int j=0;j<=i;j++){
                if(s[j]==s[i]&&(j+1>i-1||dp[j+1][i-1])){
                    dp[j][i]=1;
                    r++;
                }
            }
        }
        return r;
    }
};

int main()
{
    Solution s;
    cout<<s.countSubstrings("aaaaa")<<endl;
}
